### ISSUE TITLE
Cannot parse pydantic BaseModel as dict value since version4.29

### ISSUE THREAD

**Issue by pjgao** [omni-us/jsonargparse#555]
### To reproduce

1. Using the CLI function
The code as follows can work normally in version 4.28, but fails in versions 4.29/4.30/4.31/4.32.
```python
from typing import Dict, Optional

from jsonargparse import CLI
from pydantic import BaseModel


class InOutPuts(BaseModel):
    inputs: list[str]
    outputs: list[str]


class TrainModel(BaseModel):
    infer: Optional[Dict[str, InOutPuts]] = None


def train_func(model: TrainModel) -> None:
    print(model.infer)


if __name__ == '__main__':
    CLI(train_func)

```
2. given yaml config
```
model:
  infer:
    infer_model:
      inputs: [ 'input1','input2' ]
      outputs: [ 'logit1','logit2' ]
```
report error as follows:

```bash
error: Validation failed: Parser key "model.infer":
  Does not validate against any of the Union subtypes
  Subtypes: (typing.Dict[str, hirec.model.model_json.InOutPuts], <class 'NoneType'>)
  Errors:
    - No action for key "infer_model" to set its default.
    - Expected a <class 'NoneType'>
  Given value type: <class 'dict'>
  Given value: {'infer_model': Namespace(inputs=['id'], outputs=['dcn_logit'])}
```

### Expected behavior
parse config without error

### Environment

- jsonargparse version:4.32.0
- Python version:3.10.8
- How jsonargparse was installed: pip install jsonargparse[all]
- OS:windows11 & centos8

**Comment by mauvilsa** [comment]
Thank you for reporting! I will look at this.

**Comment by mauvilsa** [comment]
This should be fixed with #559. You are welcome to try it out. If something still doesn't work please comment.
